examples/leasing/node_release_script.py

```diff
@@ -0,0 +1,187 @@
+    return nodes
+
+
+def release_from_project(
+        hil_client, node, project
+        ):
+    """
+    We need to disconnect all the networks from all of the node's NICs
+    before releasing the node, after that we can detach node from
+    the project
+    """
+
+    node_info = hil_client.node.show(node)
+
+    for nic in node_info['nics']:
+        port = nic['port']
+        switch = nic['switch']
+        if port and switch:
+            try:
+                hil_client.port.port_revert(switch, port)
+                print('Removed all networks from node `%s`' % node)
+            except FailedAPICallException:
+                print('Failed to revert port `%s` on node \
+                        `%s` switch `%s`' % (port, node, switch))
+                raise HILClientFailure()
+
+    time.sleep(5)
+    # tries 2 times to detach the project because there might be a pending
+    # networking action setup (revert port in the previous step).
+    counter = 2
+    while counter:
+
+        try:
+            hil_client.project.detach(project, node)
+            print('Node `%s` removed from project `%s`' % (node, project))
+            break
+        except FailedAPICallException as ex:
+            if ex.message == 'Node has pending network actions':
+                counter -= 1
+                time.sleep(2)
+            else:
+                print('HIL reservation failure: Unable to \
+                        detach node `%s` from project `%s`' % (node, project))
+                raise HILClientFailure(ex.message)
+    if counter == 0:
+        print('HIL reservation failure: Unable to detach node \
+                `%s` from project `%s`' % (node, project))
+        raise HILClientFailure()
+
+
+def update_file(statusfile, nodes):
+    """
+    Writing the updated data structure's info back to the file.
+    """
+
+    with open(statusfile, 'w') as status_file:
+        for node in sorted(nodes):
+            status_file.write(str(node) + ' ' + str(nodes[node]['project']) +
+                              ' ' + str(nodes[node]['time']) + '\n')
+
+
+def release_nodes(
+        statusfile, non_persistent_list,
+        threshold_time, hil_username, hil_password, hil_endpoint,
+        hil_client=None
+        ):
+
+    """Release nodes : After certain amount of time (threshold),
+    we should return the nodes back to the free pool.
+    Tenants who need more time should ask time extension.
+    """
+
+    if not hil_client:
+        hil_client = hil_client_connect(
+                        hil_endpoint, hil_username, hil_password)
+
+    free_node_list = hil_client.node.list('free')
+    # Only these nodes should be updated in
+    # the file(either for project or for time)
+    nodes_to_update = list(set(
+        non_persistent_list)-set(free_node_list))
+
+    # All nodes should have information in status file
+    nodes = load_node_info(statusfile)
+
+    for node in nodes_to_update:
+        if nodes[node] is None:
```

Comment:
I don't think this is possible?

Reply:
This one covers the cases that someone forgets to list "all" the nodes information in the status file.

Reply:
Ah. If you want to check if the node is absent from the dictionary, do `node not in nodes`. As written, this will just raise a `KeyError`.